substrate/client/transaction-pool/src/fork_aware_txpool/txmempool.rs

```diff
@@ -0,0 +1,260 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Transaction memory pool, container for watched and unwatched transactions.
+//! Acts as a buffer which collect transactions before importing them to the views. Following are
+//! the crucial use cases when it is needed:
+//! - empty pool (no views yet)
+//! - potential races between creation of view and submitting transaction (w/o intermediary buffer
+//!   some transactions
+//! could be lost)
+//! - on some forks transaction can be invalid (view does not contain it), on other for tx can be
+//!   valid.
+
+use super::{metrics::MetricsLink as PrometheusMetrics, multi_view_listener::MultiViewListener};
+use crate::{graph, graph::ExtrinsicHash, log_xt_debug, LOG_TARGET};
+use futures::FutureExt;
+use itertools::Itertools;
+use parking_lot::RwLock;
+use sc_transaction_pool_api::TransactionSource;
+use sp_blockchain::HashAndNumber;
+use sp_runtime::{
+	traits::Block as BlockT,
+	transaction_validity::{InvalidTransaction, TransactionValidityError},
+};
+use std::{
+	collections::HashMap,
+	sync::{atomic, atomic::AtomicU64, Arc},
+	time::Instant,
+};
+
+/// The minimum interval between single transaction revalidations. Given in blocks.
+const TXMEMPOOL_REVALIDATION_PERIOD: u64 = 10;
+
+/// The number of transactions revalidated in single revalidation batch.
+const TXMEMPOOL_MAX_REVALIDATION_BATCH_SIZE: usize = 1000;
+
+/// Represents the transaction in the intermediary buffer.
+#[derive(Debug)]
+pub(crate) struct TxInMemPool<Block>
+where
+	Block: BlockT,
+{
+	//todo: add listener? for updating view with invalid transaction?
+	/// is transaction watched
+	watched: bool,
+	//todo: Arc?
```

Comment:
See https://github.com/paritytech/polkadot-sdk/pull/4639/commits/7dcacedfb286a6e2d95427d56c80cf7cca0fe695 and https://github.com/paritytech/polkadot-sdk/pull/4639/commits/469d303acb9f06be5a66c8486fac57e36a93618a.

In particular this: https://github.com/paritytech/polkadot-sdk/blob/469d303acb9f06be5a66c8486fac57e36a93618a/substrate/client/transaction-pool/src/graph/pool.rs#L48

To avoid cloning extrinsics data between views and mempool we need to have `Arc`.